src/plugins/ui_feature/find_and_replace.ts

```diff
@@ -138,35 +150,93 @@ export class FindAndReplacePlugin extends UIPlugin {
+        sheets = [
+          sheets[activeSheetIndex],
+          ...sheets.slice(activeSheetIndex + 1),
+          ...sheets.slice(0, activeSheetIndex),
+        ];
+        break;
+      case "thisSheet":
+        sheets = [this.getters.getActiveSheetId()];
+        break;
+      case "specificRange":
+        const specificRange = this.searchOptions.specificRange;
+        if (!specificRange) {
+          return [];
+        }
+        sheets = [specificRange.sheetId];
+        break;
+    }
+    return sheets;
+  }
+
   /**
    * Find matches using the current regex
    */
   private findMatches() {
-    const sheetId = this.getters.getActiveSheetId();
-    const cells = this.getters.getCells(sheetId);
     const matches: SearchMatch[] = [];
+    let sheets: UID[];
+    const specificRange = this.searchOptions.specificRange;
+    if (!this.toSearch) {
+      return matches;
+    }
+
+    sheets = this.getSearchSheets();
+
+    for (const sheetId of sheets) {
+      matches.push(...this.findMatchesInSheet(sheetId, specificRange));
+    }
+    matches.sort((a, b) => this.sortBySheetinitialStateIndexThenRowThenColumn(a, b));
+
+    const activeMatchIndex = matches.findIndex(
+      (match) => match.sheetId === this.getters.getActiveSheetId()
+    );
+    this.selectedMatchIndex = null;
+    this.selectedMatchOffset = activeMatchIndex;
+
+    return matches;
+  }
+
+  private findMatchesInSheet(sheetId: string, specificRange?: Range) {
+    const matches: SearchMatch[] = [];
+    const rangeData =
+      this.searchOptions.searchScope === "specificRange" && specificRange
+        ? specificRange.zone
+        : this.getters.getSheetZone(sheetId);
+
+    if (!this.currentSearchRegex) {
+      return [];
+    }
 
-    if (this.toSearch) {
-      for (const cell of Object.values(cells)) {
-        const { col, row } = this.getters.getCellPosition(cell.id);
-        if (
-          cell &&
-          this.currentSearchRegex &&
-          this.currentSearchRegex.test(this.getSearchableString({ sheetId, col, row }))
-        ) {
-          const match: SearchMatch = { col, row, selected: false };
+    for (let row = rangeData.top; row <= rangeData.bottom; row++) {
+      for (let col = rangeData.left; col <= rangeData.right; col++) {
+        const position: CellPosition = { sheetId, col, row };
+        if (this.currentSearchRegex.test(this.getSearchableString(position))) {
+          const match: SearchMatch = { sheetId, col, row, selected: false };
           matches.push(match);
         }
       }
     }
-    return matches.sort(this.sortByRowThenColumn);
+
+    return matches;
   }
 
-  private sortByRowThenColumn(a: Position, b: Position) {
-    if (a.row === b.row) {
-      return a.col - b.col;
-    }
-    return a.row > b.row ? 1 : -1;
+  private sortBySheetinitialStateIndexThenRowThenColumn(a: SearchMatch, b: SearchMatch) {
+    return a.sheetId !== b.sheetId
```

Comment:
Yes, you are absolutely right. In our case, the search results are already ordered. However, we also need to know the index of each search result. To achieve this, if we have the search sequence from Sheet 1, we can determine the index of each search result accurately.

Let's consider a scenario where the search scope is set to all sheets, and we are currently on Sheet 2. If we find something that already exists on Sheet 1, our functionality will start searching from Sheet 2. In this situation, if we don't have the search sequence from Sheet 1, we won't be able to understand the search index accurately. Therefore, we require this function to obtain the search sequence from Sheet 1.

Reply:
Ok, I think I understand the issue. If I'm not mistaken, the search index is computed in the plugin such that it would be correctly displayed in the sidepanel. I don't think it's good or bad in itself but since it's important for the sidepanel*, there should be an explicitely related test. I don't think there is because when I removed the sorting function, the side panel tests did not fail.

So, I think that the function findMatches just does too many things. it should not change the state of the plugin just yield the results, regardless of the current sheet etc. 
the logic of the offset should be handled on the fly in `selectNextCell` when `selectedMatchIndex=== null`, instead of relying on a variable computed elsewhere. Otherwise, you have 2 functions but one of them needs to be called before the second one for the code to make any sense.

What I would do in `refreshSearch`:
- Before calling findMatches, reset selectedMatchIndex
- this.searchMatches = this.findMatches()
At this point we have properly reset the state according to the new search regex
- `this.findMatches` only finds the matches by looping over all sheets starting with the first one
- then call SelectNextCell (which probably should be called `selectNextMatch` btw) and in that function IIF `selectedMatchIndex === null` then you compute the offset

Reply:
Hello @rrahir, could you please review the changes? I have added the test for this and made the necessary adjustments accordingly.